File: Data Structures & Algorithms/balanced-binary-tree/submission-4.js
/**
 * Definition for a binary tree node.
 * class TreeNode {
 *     constructor(val = 0, left = null, right = null) {
 *         this.val = val;
 *         this.left = left;
 *         this.right = right;
 *     }
 * }
 */

class Solution {
    /**
     * @param {TreeNode} root
     * @return {boolean}
     */
    isBalanced(root) {
        let isBalanced = true;

        function maxDepthDFS(root) {
            if (!root) return 0;
            const left = maxDepthDFS(root.left);
            const right = maxDepthDFS(root.right);
            isBalanced &&= (Math.abs(left - right) <= 1);
            return Math.max(left, right) + 1;
        }

        maxDepthDFS(root);
        return isBalanced;
    }
}

File: Data Structures & Algorithms/binary-tree-maximum-path-sum/submission-4.cpp
/**
 * Definition for a binary tree node.
 * struct TreeNode {
 *     int val;
 *     TreeNode *left;
 *     TreeNode *right;
 *     TreeNode() : val(0), left(nullptr), right(nullptr) {}
 *     TreeNode(int x) : val(x), left(nullptr), right(nullptr) {}
 *     TreeNode(int x, TreeNode *left, TreeNode *right) : val(x), left(left), right(right) {}
 * };
 */

class Solution {
private:
    int max_path = INT_MIN;
public:
    int maxPathSum(TreeNode* root) {
        maxGainFrom(root);
        return max_path;
    }
    int maxGainFrom(TreeNode* root) {
        if (!root) return 0;
        int left_max = max(maxGainFrom(root->left), 0);
        int right_max = max(maxGainFrom(root->right), 0);
        max_path = max(max_path, left_max + right_max + root->val);
        return root->val + max(left_max, right_max);
    }
};
